### PR TITLE
Add initial client cert support and ssl server changes

### PR DESCRIPTION
**Pull request by Jotschi**

### CODE REVIEW

File: api/src/main/java/com/gentics/mesh/etc/config/HttpServerConfig.java [hunk 1]
```diff
@@ -23,15 +28,21 @@
 
 	public static final int DEFAULT_HTTP_PORT = 8080;
 
+	public static final int DEFAULT_HTTPS_PORT = 8443;
+
 	public static final String DEFAULT_CERT_PATH = "config/cert.pem";
 	public static final String DEFAULT_KEY_PATH = "config/key.pem";
+	public static final ClientAuth DEFAULT_CLIENT_AUTH_MODE = ClientAuth.REQUEST;
```
Comment (yrucrem) [inline api/src/main/java/com/gentics/mesh/etc/config/HttpServerConfig.java]: Are the SSL handshakes  more concise when this is set to `ClientAuth.NONE`? If so maybe rethink the default value.

**Comment by Jotschi** [comment]
TODO:
* Check whether having multiple servers in one verticle could cause issues
* Check whether clientCertMode request would cause additional ssl overhead
* Check whether it is really needed to add the server cert to the trusted cert list
* Add tests
* Add changelog

File: api/src/main/java/com/gentics/mesh/etc/config/HttpServerConfig.java [hunk 2]
```diff
@@ -78,6 +99,13 @@
 	@EnvironmentVariable(name = MESH_HTTP_SSL_KEY_PATH_ENV, description = "Override the configured SSL enable flag.")
 	private String keyPath = DEFAULT_KEY_PATH;
 
+	@JsonProperty(required = false)
+	@JsonPropertyDescription("Configure the client certificate handling mode. Options: none, request, required. Default: none")
```
Comment (Jotschi) [inline api/src/main/java/com/gentics/mesh/etc/config/HttpServerConfig.java]: Fix doc (default).

Comment (yrucrem) [inline api/src/main/java/com/gentics/mesh/etc/config/HttpServerConfig.java]: The default value is actually `ClientAuth.REQUEST`.

File: core/src/main/java/com/gentics/mesh/rest/RestAPIVerticle.java [hunk 3]
```diff
@@ -155,50 +161,64 @@ public RestAPIVerticle() {
...
-		HttpServerOptions options = new HttpServerOptions();
-		options.setPort(port);
-		options.setHost(host);
-		options.setCompressionSupported(true);
-		options.setHandle100ContinueAutomatically(true);
+		HttpServerConfig meshServerOptions = meshOptions.getHttpServerOptions();
+		if (meshServerOptions.isHttp()) {
+			HttpServerOptions httpOptions = new HttpServerOptions();
+			if (log.isDebugEnabled()) {
+				log.debug("Setting ssl server options..");
```
Comment (yrucrem) [inline core/src/main/java/com/gentics/mesh/rest/RestAPIVerticle.java]: Remove this log message (this is about HTTP)

**Comment by Jotschi** [comment]
https://github.com/gentics/mesh/issues/991

File: api/src/main/java/com/gentics/mesh/etc/config/env/OptionUtils.java [hunk 4]
```diff
@@ -44,6 +47,12 @@
 		} else if (clazz.isEnum()) {
 			Object enumValue = Enum.valueOf((Class<Enum>)clazz, value);
 			return (T) enumValue;
+		} else if(clazz.equals(List.class)) {
+			if (value == null || value.trim().length() == 0) {
+				return (T) new ArrayList<String>();
```
Comment (philippguertler) [inline api/src/main/java/com/gentics/mesh/etc/config/env/OptionUtils.java]:
```suggestion
				return Collections.emptyList();
```

File: core/src/main/java/com/gentics/mesh/rest/RestAPIVerticle.java [hunk 5]
```diff
@@ -60,7 +65,9 @@
 
 	private static final Logger log = LoggerFactory.getLogger(AbstractInternalEndpoint.class);
 
-	protected HttpServer server;
+	protected HttpServer httpsServer;
```
Comment (philippguertler) [inline core/src/main/java/com/gentics/mesh/rest/RestAPIVerticle.java]:
```suggestion
	private HttpServer httpsServer;
```

File: core/src/main/java/com/gentics/mesh/rest/RestAPIVerticle.java [hunk 6]
```diff
@@ -60,7 +65,9 @@
...
+
+	protected HttpServer httpServer;
```
Comment (philippguertler) [inline core/src/main/java/com/gentics/mesh/rest/RestAPIVerticle.java]:
```suggestion
	private HttpServer httpServer;
```

File: core/src/main/java/com/gentics/mesh/rest/RestAPIVerticle.java [hunk 7]
```diff
@@ -155,50 +165,62 @@ public RestAPIVerticle() {
...
-				promise.fail("Could not find SSL key within path {" + httpServerOptions.getKeyPath() + "}");
+			if (!Paths.get(meshServerOptions.getKeyPath()).toFile().exists()) {
+				promise.fail("Could not find SSL key within path {" + meshServerOptions.getKeyPath() + "}");
 				return;
 			}
-			if (!Paths.get(httpServerOptions.getCertPath()).toFile().exists()) {
-				promise.fail("Could not find SSL cert within path {" + httpServerOptions.getCertPath() + "}");
+			if (!Paths.get(meshServerOptions.getCertPath()).toFile().exists()) {
+				promise.fail("Could not find SSL cert within path {" + meshServerOptions.getCertPath() + "}");
 				return;
```
Comment (philippguertler) [inline core/src/main/java/com/gentics/mesh/rest/RestAPIVerticle.java]: Move this to `HttpServerConfig#validate`

File: core/src/main/java/com/gentics/mesh/rest/RestAPIVerticle.java [hunk 8]
```diff
@@ -155,50 +165,62 @@ public RestAPIVerticle() {
...
+
+			keyOptions.setKeyPath(meshServerOptions.getKeyPath());
+			keyOptions.setCertPath(meshServerOptions.getCertPath());
+			httpsOptions.setPemKeyCertOptions(keyOptions);
+
+			PemTrustOptions pemTrustOptions = new PemTrustOptions();
+			for (String path : meshServerOptions.getTrustedCertPaths()) {
+				pemTrustOptions.addCertPath(path);
+			}
+			// Finally also add the server cert
```
Comment (philippguertler) [inline core/src/main/java/com/gentics/mesh/rest/RestAPIVerticle.java]: Check if this works without adding trusted certs

File: core/src/main/java/com/gentics/mesh/rest/RestAPIVerticle.java [hunk 9]
```diff
@@ -209,34 +231,46 @@ public void start(Promise<Void> promise) throws Exception {
...
+		List<Single<HttpServer>> serverListens = Arrays.asList(httpServer, httpsServer).stream()
+			.filter(Objects::nonNull)
+			.map(s -> {
+				return s.requestHandler(rxRootRouter);
+			})
+			.map(s -> s.rxListen())
+			.collect(Collectors.toList());
+
+		Single.merge(serverListens).ignoreElements().subscribe(() -> {
+			promise.complete();
```
Comment (philippguertler) [inline core/src/main/java/com/gentics/mesh/rest/RestAPIVerticle.java]: Use method references

File: core/src/main/java/com/gentics/mesh/rest/RestAPIVerticle.java [hunk 10]
```diff
@@ -287,8 +321,12 @@ private void registerEndPoints(RouterStorage storage) throws Exception {
...
 
-	public HttpServer getServer() {
-		return server;
+	public HttpServer getHttpServer() {
+		return httpServer;
+	}
+
+	public HttpServer getHttpsServer() {
+		return httpsServer;
 	}
```
Comment (philippguertler) [inline core/src/main/java/com/gentics/mesh/rest/RestAPIVerticle.java]: Remove getters

File: core/src/test/java/com/gentics/mesh/core/ssl/SSLTestClient.java [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+		char[] randomKeyStorePass = UUIDUtil.randomUUID().toCharArray();
+		CertificateFactory certificateFactory = CertificateFactory.getInstance("X.509");
+		X509Certificate clientCert = (X509Certificate) certificateFactory.generateCertificate(new FileInputStream(CLIENT_CERT_PEM));
+		KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+		KeyManagerFactory keyManagerFactory = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+		String principal = clientCert.getSubjectX500Principal().getName();
+
+		try (PEMParser pemParser = new PEMParser(new FileReader(CLIENT_KEY_PEM))) {
+			PrivateKey privateKey = new JcaPEMKeyConverter().getPrivateKey((PrivateKeyInfo) pemParser.readObject());
+			keyStore.load(null);
```
Comment (philippguertler) [inline core/src/test/java/com/gentics/mesh/core/ssl/SSLTestClient.java]: What is the purpose of this?

File: rest-client/src/main/java/com/gentics/mesh/rest/client/MeshRestClientConfig.java [hunk 12]
```diff
@@ -144,5 +193,95 @@ public Builder setBasePath(String basePath) {
...
+			return this;
+		}
+
+		/**
+		 * Set the InputStream from which the client SSL cert in PEM format will be read.
+		 * 
+		 * @param ins
+		 * @return
+		 */
+		public Builder setClientCert(InputStream ins) {
```
Comment (philippguertler) [inline rest-client/src/main/java/com/gentics/mesh/rest/client/MeshRestClientConfig.java]: Add setter with `byte[]`

File: rest-client/src/main/java/com/gentics/mesh/rest/client/MeshRestClientConfig.java [hunk 13]
```diff
@@ -144,5 +193,95 @@ public Builder setBasePath(String basePath) {
...
+			return this;
+		}
+
+		/**
+		 * Set the filesystem path to the CA file which is formatted in PEM format.
+		 * 
+		 * @param path
+		 * @return
+		 */
+		public Builder setTrustedCA(String path) {
```
Comment (philippguertler) [inline rest-client/src/main/java/com/gentics/mesh/rest/client/MeshRestClientConfig.java]: Allow multiple CAs or additional CAs

File: rest-client/src/main/java/com/gentics/mesh/rest/client/impl/MeshRestOkHttpClientImpl.java [hunk 14]
```diff
@@ -33,23 +31,15 @@ public MeshRestOkHttpClientImpl(MeshRestClientConfig config, OkHttpClient client
...
 	/**
-	 * We need a long timeout per default since some requests take a long time. For all tests a 1 minute timeout works fine.
+	 * Return the default OkHttpClient
 	 * 
+	 * @param MeshRestClientConfig
+	 *            config
 	 * @return
 	 */
-	private static OkHttpClient defaultClient() {
+	private static OkHttpClient defaultClient(MeshRestClientConfig config) {
```
Comment (philippguertler) [inline rest-client/src/main/java/com/gentics/mesh/rest/client/impl/MeshRestOkHttpClientImpl.java]: Remove static default client

File: core/src/test/java/com/gentics/mesh/core/ssl/SSLTestClient.java (hunk 11 above)
Comment (Jotschi) [reply core/src/test/java/com/gentics/mesh/core/ssl/SSLTestClient.java]: Those are the loading parameters. There is no method without any parameters. Thus this option is used. JavaDoc of `load` allows null in this case.

File: core/src/test/java/com/gentics/mesh/core/ssl/SSLTestClient.java [hunk 15]
```diff
@@ -1,159 +1,58 @@
...
 import io.vertx.core.logging.Logger;
 import io.vertx.core.logging.LoggerFactory;
 import okhttp3.OkHttpClient;
-import okhttp3.OkHttpClient.Builder;
 import okhttp3.Request;
 import okhttp3.Response;
 
 public class SSLTestClient {
 
+	public enum CLIENT_CERT {
```
Comment (philippguertler) [inline core/src/test/java/com/gentics/mesh/core/ssl/SSLTestClient.java]: Rename to ClientCert

**Pull request merged**